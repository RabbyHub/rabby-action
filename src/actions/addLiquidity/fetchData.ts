import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
