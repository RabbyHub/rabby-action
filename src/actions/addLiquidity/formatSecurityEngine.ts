import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
