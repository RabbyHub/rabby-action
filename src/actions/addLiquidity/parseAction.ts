import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
