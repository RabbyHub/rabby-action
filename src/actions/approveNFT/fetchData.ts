import PQueue from 'p-queue';
import { ApproveNFTRequireData, FetchActionRequiredData } from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

export const fetchDataApproveNFT: FetchActionRequiredData<{
  spender: string;
}> = async (options, likeAction) => {
  const queue = new PQueue();
  const { sender, apiProvider, chainId, actionData, extraActionDataState } =
    options;
  const action = likeAction || actionData.approveNFT;

  if (!action) {
    return {};
  }

  const result: ApproveNFTRequireData = {
    isEOA: false,
    contract: null,
    riskExposure: 0,
    rank: null,
    bornAt: 0,
    protocol: null,
    isDanger: false,
    tokenBalance: '0',
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(
      action.spender,
      chainId
    );
    if (!contractInfo) {
      result.isEOA = true;
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.riskExposure = contractInfo.top_nft_spend_usd_value;
      result.bornAt = contractInfo.create_at;
      result.isDanger =
        contractInfo.is_danger.auto || contractInfo.is_danger.edit;
      result.protocol = contractInfo.protocol;
    }

    if (result.isEOA) {
      queue.add(async () => {
        const { desc } = await apiProvider.addrDesc(action.spender);
        result.bornAt = desc.born_at;
      });
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: action.spender,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
