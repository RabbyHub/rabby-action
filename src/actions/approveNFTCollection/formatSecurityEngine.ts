import {
  ApproveNFTRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineApproveNFTCollection: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId, provider } = options;

    if (!actionData.approveNFTCollection || !chainId) {
      return {};
    }
    const data = requireData as ApproveNFTRequireData;
    const { spender } = actionData.approveNFTCollection;
    return {
      collectionApprove: {
        chainId,
        spender,
        isEOA: data.isEOA,
        riskExposure: data.riskExposure,
        deployDays: provider.getTimeSpan(
          Math.floor(Date.now() / 1000) - data.bornAt
        ).d,
        isDanger: !!data.isDanger,
      },
    };
  };
