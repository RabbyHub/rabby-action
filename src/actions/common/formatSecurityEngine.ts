import { ContractRequireData, FormatSecurityEngineContext } from '../../types';

export const formatSecurityEngineCommon: FormatSecurityEngineContext = async (
  options
) => {
  if (!options.actionData.common) {
    return {};
  }

  return {
    common: {
      ...options.actionData.common,
      receiverInWallet: (options.requireData as ContractRequireData)
        ?.receiverInWallet,
    },
  };
};
