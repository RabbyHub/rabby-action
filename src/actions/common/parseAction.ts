import { ParseAction } from '../../types';

export const parseActionCommon: ParseAction = (options) => {
  const { data } = options;
  if (data?.type !== null) {
    return {};
  }

  const from =
    options.type === 'transaction'
      ? options.sender
      : options.type === 'text'
      ? undefined
      : options.sender;

  return {
    common: {
      from,
      ...(data as any),
    },
  };
};
