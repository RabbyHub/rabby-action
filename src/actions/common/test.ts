import {
  apiProvider,
  formatProvider,
  walletProvider,
  SENDER,
  ETH_CHAIN_ID,
  ORIGIN,
} from '../../../__mocks__';
import {
  parseTxDataTypedData,
  parseTxDataTypedDataWithAction,
  txDataTypedData,
  txDataTypedDataWithAction,
} from './mocks';
import { parseActionCommon } from './parseAction';
import { fetchDataCommon } from './fetchData';
import { formatSecurityEngineCommon } from './formatSecurityEngine';
import { parseTypedDataAction } from '../../utils/parseTypedDataAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * https://metamask.github.io/test-dapp/#signTypedDataV4
 * Sign Typed Data V4
 * [Sign] button
 */
test.each([
  [
    parseTypedDataAction(parseActionCommon),
    fetchDataCommon,
    formatSecurityEngineCommon,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('Common -> TypedData', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'typed_data',
    data: parseTxDataTypedData['action'],
    typedData: txDataTypedData,
    sender: SENDER,
  });
  expect(actionData).toMatchSnapshot('parseActionCommon');

  const requireData = await _fetchData({
    type: 'typed_data',
    actionData,
    chainId: ETH_CHAIN_ID,
    walletProvider,
    apiProvider,
    sender: SENDER,
  });
  expect(requireData).toMatchSnapshot('fetchDataCommon');

  const ctx = await _format({
    type: 'typed_data',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
    origin: ORIGIN,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineCommon');
});

/**
 * send
 *  ```json
    {
      "method": "eth_signTypedData_v4",
      "params": [
        "0x341a1fbd51825e5a107db54ccb3166deba145479",
        "{\"types\":{\"ClobAuth\":[{\"name\":\"address\",\"type\":\"address\"},{\"name\":\"timestamp\",\"type\":\"string\"},{\"name\":\"nonce\",\"type\":\"uint256\"},{\"name\":\"message\",\"type\":\"string\"}],\"EIP712Domain\":[{\"name\":\"name\",\"type\":\"string\"},{\"name\":\"version\",\"type\":\"string\"},{\"name\":\"chainId\",\"type\":\"uint256\"}]},\"domain\":{\"name\":\"ClobAuthDomain\",\"version\":\"1\",\"chainId\":\"137\"},\"primaryType\":\"ClobAuth\",\"message\":{\"address\":\"0x341a1fbd51825e5a107db54ccb3166deba145479\",\"timestamp\":\"1724848671\",\"nonce\":\"0\",\"message\":\"This message attests that I control the given wallet\"}}"
      ]
    }
  ```
 */
test.each([
  [
    parseTypedDataAction(parseActionCommon),
    fetchDataCommon,
    formatSecurityEngineCommon,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])(
  'Common -> TypedData -> with action',
  async (_parseAction, _fetchData, _format) => {
    const actionData = _parseAction({
      type: 'typed_data',
      data: parseTxDataTypedDataWithAction['action'],
      typedData: txDataTypedDataWithAction,
      sender: SENDER,
    });
    expect(actionData).toMatchSnapshot('parseActionCommon');

    const requireData = await _fetchData({
      type: 'typed_data',
      actionData,
      chainId: ETH_CHAIN_ID,
      walletProvider,
      apiProvider,
      sender: SENDER,
    });
    expect(requireData).toMatchSnapshot('fetchDataCommon');

    const ctx = await _format({
      type: 'typed_data',
      actionData,
      requireData,
      chainId: ETH_CHAIN_ID,
      isTestnet: false,
      provider: formatProvider,
      origin: ORIGIN,
    });
    expect(ctx).toMatchSnapshot('formatSecurityEngineCommon');
  }
);
