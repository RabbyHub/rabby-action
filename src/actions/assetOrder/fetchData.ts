import { FetchActionRequiredData } from '../../types';
import { fetchDataCommon } from '../common/fetchData';

export const fetchDataAssetOrder: FetchActionRequiredData = async (options) => {
  let action;

  if (options.type === 'transaction' && options.actionData.assetOrder) {
    action = {
      receiver: options.tx.to,
    };
  }
  if (
    options.type === 'typed_data' &&
    options.actionData.assetOrder &&
    options.actionData.contractId
  ) {
    action = {
      receiver: options.actionData.contractId,
    };
  }

  if (!action || !options.chainId) {
    return {};
  }

  return fetchDataCommon(options, action);
};
