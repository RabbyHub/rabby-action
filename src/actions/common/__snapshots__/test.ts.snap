// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common -> TypedData -> with action: fetchDataCommon 1`] = `{}`;

exports[`Common -> TypedData -> with action: fetchDataCommon 2`] = `{}`;

exports[`Common -> TypedData -> with action: formatSecurityEngineCommon 1`] = `
{
  "common": {
    "desc": "Authorization",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "is_asset_changed": false,
    "is_involving_privacy": true,
    "receiverInWallet": undefined,
    "title": "Authorization",
    "type": null,
  },
}
`;

exports[`Common -> TypedData -> with action: formatSecurityEngineCommon 2`] = `
{
  "common": {
    "desc": "Authorization",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "is_asset_changed": false,
    "is_involving_privacy": true,
    "receiverInWallet": undefined,
    "title": "Authorization",
    "type": null,
  },
}
`;

exports[`Common -> TypedData -> with action: parseActionCommon 1`] = `
{
  "actionType": null,
  "brand": undefined,
  "chainId": "137",
  "common": {
    "desc": "Authorization",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "is_asset_changed": false,
    "is_involving_privacy": true,
    "title": "Authorization",
    "type": null,
  },
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`Common -> TypedData -> with action: parseActionCommon 2`] = `
{
  "actionType": null,
  "brand": undefined,
  "chainId": "137",
  "common": {
    "desc": "Authorization",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "is_asset_changed": false,
    "is_involving_privacy": true,
    "title": "Authorization",
    "type": null,
  },
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`Common -> TypedData: fetchDataCommon 1`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
  },
  "hasInteraction": true,
  "id": "0xcccccccccccccccccccccccccccccccccccccccc",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "unexpectedAddr": null,
}
`;

exports[`Common -> TypedData: fetchDataCommon 2`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
  },
  "hasInteraction": true,
  "id": "0xcccccccccccccccccccccccccccccccccccccccc",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "unexpectedAddr": null,
}
`;

exports[`Common -> TypedData: formatSecurityEngineCommon 1`] = `{}`;

exports[`Common -> TypedData: formatSecurityEngineCommon 2`] = `
{
  "contractCall": {
    "chainId": "eth",
    "id": "0xcccccccccccccccccccccccccccccccccccccccc",
  },
}
`;

exports[`Common -> TypedData: parseActionCommon 1`] = `
{
  "actionType": null,
  "brand": undefined,
  "chainId": 1,
  "contractId": "0xcccccccccccccccccccccccccccccccccccccccc",
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`Common -> TypedData: parseActionCommon 2`] = `
{
  "actionType": null,
  "brand": undefined,
  "chainId": 1,
  "contractCall": {},
  "contractId": "0xcccccccccccccccccccccccccccccccccccccccc",
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;
